compactor2/src/components/partition_filter/max_num_columns.rs

```diff
@@ -0,0 +1,105 @@
+use std::fmt::Display;
+
+use async_trait::async_trait;
+use data_types::ParquetFile;
+
+use crate::{
+    error::{DynError, ErrorKind, SimpleError},
+    PartitionInfo,
+};
+
+use super::PartitionFilter;
+
+#[derive(Debug)]
+pub struct MaxNumColumnsPartitionFilter {
+    max_num_columns: usize,
+}
+
+impl MaxNumColumnsPartitionFilter {
+    pub fn new(max_num_columns: usize) -> Self {
+        Self { max_num_columns }
+    }
+}
+
+impl Display for MaxNumColumnsPartitionFilter {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "max_num_columns")
+    }
+}
+
+#[async_trait]
+impl PartitionFilter for MaxNumColumnsPartitionFilter {
+    async fn apply(
+        &self,
+        partition_info: &PartitionInfo,
+        _files: &[ParquetFile],
```

Comment:
Instead of changing the first argument to pass `PartitionInfo`, I think it would be nicer to check the file metadata itself! So swap this back to:

```suggestion
        _partition_id: PartitionId,
        files: &[ParquetFile],
```

And then use `files.column_set` to inspect the number of columns in each file. 

A merge operation between two `ColumnSet` should be created that performs a set union (deduplicating the columns) and that called on all `column_set` in `files` to produce the final set of columns across all files. Then a simple `final_columns.len()` will tell you how many columns there are in this compaction job, and we can enforce it cheaply without any extra queries 🎉

Reply:
I can help with the `ColumnSet` work if you want a hand @NGA-TRAN  :+1: